flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/CancelingTest.java

```diff
@@ -0,0 +1,171 @@
+
+            canceling.onEnter();
+            assertThat(cancellableExecutionGraph.isCancelled(), is(true));
+        }
+    }
+
+    @Test
+    public void testTransitionToFinishedWhenCancellationCompletes() throws Exception {
+        try (MockStateWithExecutionGraphContext ctx = new MockStateWithExecutionGraphContext()) {
+            Canceling canceling = createCancelingState(ctx);
+
+            ctx.setExpectFinished(
+                    archivedExecutionGraph ->
+                            assertThat(archivedExecutionGraph.getState(), is(JobStatus.CANCELED)));
+            canceling.getExecutionGraph().cancel(); // this calls onTerminalState.onTerminalState()
+        }
+    }
+
+    @Test
+    public void testTransitionToSuspend() throws Exception {
+        try (MockStateWithExecutionGraphContext ctx = new MockStateWithExecutionGraphContext()) {
+            Canceling canceling = createCancelingState(ctx);
+
+            ctx.setExpectFinished(
+                    archivedExecutionGraph ->
+                            assertThat(archivedExecutionGraph.getState(), is(JobStatus.SUSPENDED)));
+            canceling.suspend(new RuntimeException("suspend"));
+        }
+    }
+
+    @Test
+    public void testCancelIsIgnored() throws Exception {
+        try (MockStateWithExecutionGraphContext ctx = new MockStateWithExecutionGraphContext()) {
+            Canceling canceling = createCancelingState(ctx);
+            canceling.cancel();
+            ctx.assertNoStateTransition();
+        }
+    }
+
+    @Test
+    public void testGlobalFailuresAreIgnored() throws Exception {
+        try (MockStateWithExecutionGraphContext ctx = new MockStateWithExecutionGraphContext()) {
+            Canceling canceling = createCancelingState(ctx);
+            canceling.handleGlobalFailure(new RuntimeException("test"));
+            ctx.assertNoStateTransition();
+        }
+    }
+
+    @Test
+    public void testTaskFailuresAreIgnored() throws Exception {
+        try (MockStateWithExecutionGraphContext ctx = new MockStateWithExecutionGraphContext()) {
+            Canceling canceling = createCancelingState(ctx);
+            // register execution at EG
+            ExecutingTest.MockExecutionJobVertex ejv =
+                    new ExecutingTest.MockExecutionJobVertex(canceling.getExecutionGraph());
+            TaskExecutionStateTransition update =
+                    new TaskExecutionStateTransition(
+                            new TaskExecutionState(
+                                    canceling.getJob().getJobID(),
+                                    ejv.getMockExecutionVertex()
+                                            .getCurrentExecutionAttempt()
+                                            .getAttemptId(),
+                                    ExecutionState.FAILED,
+                                    new RuntimeException()));
+            canceling.updateTaskExecutionState(update);
+            ctx.assertNoStateTransition();
```

Comment:
We should also assert that the `ExecutionGraph` has been updated. Otherwise a passing implementation of the `Canceling.updateTaskExecutionState` could be an empty method.

Technically, we would also have to do this for the other `StateWithExecutionGraph` sub classes.